### ISSUE TITLE
Removing jQuery from GitHub.com frontend

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#55]
Removing jQuery from GitHub.com frontend<br>
<br>
https://ift.tt/2Nnqr3f<br>
<br>
Mislav Marohni&#263;<br>
<br>
<p>We have recently completed a milestone where we were able to drop jQuery as a dependency of the frontend code for GitHub.com. This marks the end of a gradual, years-long transition of increasingly decoupling from jQuery until we were able to completely remove the library. In this post, we will explain a bit of history of how we started depending on jQuery in the first place, how we realized when it was no longer needed, and point out that&mdash;instead of replacing it with another library or framework&mdash;we were able to achieve everything that we needed using standard browser APIs.</p>
<h2>Why jQuery made sense early on</h2>
<p>GitHub.com pulled in <a href="https://blog.jquery.com/2007/09/16/jquery-1-2-1-released/">jQuery 1.2.1</a> as a dependency in late 2007. For a bit of context, that was a year before Google released the first version of their Chrome browser. There was no standard way to query DOM elements by a CSS selector, no standard way to animate visual styles of an element, and the <a href="https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest">XMLHttpRequest interface</a> pioneered by Internet Explorer was, like many other APIs, inconsistent between browsers.</p>
<p>jQuery made it simple to manipulate the DOM, define animations, and make &ldquo;AJAX&rdquo; requests&mdash; basically, it enabled web developers to create more modern, dynamic experiences that stood out from the rest. Most importantly of all, the JavaScript features built in one browser with jQuery would generally work in other browsers, too. In those early days of GitHub when most of its features were still getting fleshed out, this allowed the small development team to prototype rapidly and get new features out the door without having to adjust code specifically for each web browser.</p>
<p>The simple interface of jQuery also served as a blueprint to craft extension libraries that would later serve as building blocks for the rest of GitHub.com frontend: <a href="https://github.com/defunkt/jquery-pjax">pjax</a> and <a href="https://github.com/defunkt/facebox">facebox</a>.</p>
<p>We will always be thankful to John Resig and the jQuery contributors for creating and maintaining such a useful and, for the time, <em>essential</em> library.</p>
<h2>Web standards in the later years</h2>
<p>Over the years, GitHub grew into a company with hundreds of engineers and a dedicated team gradually formed to take responsibility for the size and quality of JavaScript code that we serve to web browsers. One of the things that we&rsquo;re constantly on the lookout for is technical debt, and sometimes technical debt grows around dependenices that once provided value, but whose value dropped over time.</p>
<p>When it came to jQuery, we compared it against the rapid evolution of supported web standard in modern browsers and realized:</p>
<ul>
<li>The <code>$(selector)</code> pattern can easily be replaced with <code>querySelectorAll()</code>;</li>
<li>CSS classname switching can now be achieved using <a href="https://developer.mozilla.org/en-US/docs/Web/API/Element/classList">Element.classList</a>;</li>
<li>CSS now supports defining <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Animations/Using_CSS_animations">visual animations in stylesheets</a> rather than in JavaScript;</li>
<li>
<code>$.ajax</code> requests can be performed using the <a href="https://fetch.spec.whatwg.org/">Fetch Standard</a>;</li>
<li>The <a href="https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener"><code>addEventListener()</code></a> interface is stable enough for cross-platform use;</li>
<li>We could easily encapsulate <a href="https://github.com/dgraham/delegated-events#readme">the event delegation pattern</a> with a lightweight library;</li>
<li>Some syntactic sugar that jQuery provides has become reduntant with the evolution of JavaScript language.</li>
</ul>
<p>Furthermore, the chaining syntax didn&rsquo;t satisfy how we wanted to write code going forward. For example:</p>
<div>
<div>
<pre>
<code><span>$</span><span>(</span><span>'.js-widget'</span><span>)</span>
  <span>.</span><span>addClass</span><span>(</span><span>'is-loading'</span><span>)</span>
  <span>.</span><span>show</span><span>()</span>
</code>
</pre>
</div>
</div>
<p>This syntax is simple to write, but to our standards, doesn&rsquo;t communicate intent really well. Did the author expect one or more <code>js-widget</code> elements on this page? Also, if we update our page markup and accidentally leave out the <code>js-widget</code> classname, will an exception in the browser inform us that something went wrong? By default, jQuery silently skips the whole expresion when nothing matched the initial selector; but to us, such behavior was a bug rather than a feature.</p>
<p>Finally, we wanted to start annotating types with <a href="https://flow.org/">Flow</a> to perform static type checking at build time, and we concluded that the chaining syntax doesn&rsquo;t lend itself well to static analysis, since almost every result of a jQuery method call is of the same type. We chose Flow over alternatives because, at the time, features such as <code>@flow weak</code> mode allowed us to progressively and efficiently start applying types to a codebase which was largely untyped.</p>
<p>All in all, decoupling from jQuery would mean that we could rely on web standards more, have <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript">MDN web docs</a> be de-facto default documentation for our frontend developers, maintain more resilient code in the future, and eventually drop a 30 kB dependency from our packaged bundles, speeding up page load times and JavaScript execution times.</p>
<h2>Incremental decoupling</h2>
<p>Even with an end goal in sight, we knew that it wouldn&rsquo;t be feasible to just allocate all resources we had to rewriting everything from jQuery to vanilla JS. If anything, such a rushed endeavor would likely lead to many regressions in site functionality that we would later have to weed out. Instead, we:</p>
<ul>
<li>
<p>Set up metrics that tracked ratio of jQuery calls used per overall line of code and monitored that graph over time to make sure that it&rsquo;s either staying constant or going down, not up.</p>
<p><img alt="Graph of jQuery usage going down over time." src="https://githubengineering.com/images/removing-jquery-from-github-frontend/jquery-usage.png"></p>
</li>
<li>
<p>We discouraged importing jQuery in any new code. To facilitate that using automation, we created <a href="https://github.com/dgraham/eslint-plugin-jquery#readme">eslint-plugin-jquery</a> which would fail CI checks if anyone tried to use jQuery features, for example <code>$.ajax</code>.</p>
</li>
<li>
<p>There were now plenty of violations of eslint rules in old code, all of which we&rsquo;ve annotated with specific <code>eslint-disable</code> rules in code comments. To the reader of that code, those comments would serve as a clear signal that this code doesn&rsquo;t represent our current coding practices.</p>
</li>
<li>
<p>We created a pull request bot that would leave a review comment on a pull request pinging our team whenever somebody tried to add a new <code>eslint-disable</code> rule. This way we would get involved in code review early and suggest alternatives.</p>
</li>
<li>
<p>A lot of old code had explicit coupling to external interfaces of pjax and facebox jQuery plugins, so we&rsquo;ve kept their interfaces relatively the same while we&rsquo;ve internally replaced their implementation with vanilla JS. Having static type checking helped us have greater confidence around those refactorings.</p>
</li>
<li>
<p>Plenty of old code interfaced with <a href="http://josh.github.io/rails-behaviors/">rails-behaviors</a>, our adapter for the Ruby on Rails approach to &ldquo;unobtrusive&rdquo; JS, in a way that they would attach an AJAX lifecycle handler to certain forms:</p>
<div>
<div>
<pre>
<code>  <span>// LEGACY APPROACH</span>
  <span>$</span><span>(</span><span>document</span><span>).</span><span>on</span><span>(</span><span>'ajaxSuccess'</span><span>,</span> <span>'form.js-widget'</span><span>,</span> <span>function</span><span>(</span><span>event</span><span>,</span> <span>xhr</span><span>,</span> <span>settings</span><span>,</span> <span>data</span><span>)</span> <span>{</span>
    <span>// insert response data somewhere into the DOM</span>
  <span>})</span>
</code>
</pre>
</div>
</div>
<p>Instead of having to rewrite all of those call sites at once to the new approach, we&rsquo;ve opted to trigger <em>fake</em> <code>ajax*</code> lifecycle events and keep these forms submitting their contents asynchronously as before; only this time <code>fetch()</code> was used internally.</p>
</li>
<li>
<p>We maintained a custom build of jQuery and whenever we&rsquo;ve identified that we&rsquo;re not using a certain module of jQuery anymore, we would remove it from the custom build and ship a slimmer version. For instance, after we have removed the final usage of jQuery-specific CSS pseudo-selectors such as <code>:visible</code> or <code>:checkbox</code>, we were able to remove the <a href="https://sizzlejs.com/">Sizzle module</a>; and when the last <code>$.ajax</code> call was replaced with <code>fetch()</code>, we were able to remove the AJAX module. This served a dual purpose: speeding up JavaScript execution times while at the same time ensuring that no new code is created that would try using the removed functionality.</p>
</li>
<li>
<p>We kept dropping support for old Internet Explorer versions as soon as it would be feasible to, as informed by our site analytics. Whenever use of a certain IE version dropped below a certain threshold, we would stop serving JavaScript to it and focus on testing against and supporting more modern browsers. Dropping support for IE 8&ndash;9 early on allowed us to adopt many native browser features that would otherwise be hard to polyfill.</p>
</li>
<li>
<p>As part of our refined approach to building frontend features on GitHub.com, we focused on getting away with regular HTML foundation as much as we could, and only adding JavaScript behaviors as progressive enhancement. As a result, even those web forms and other UI elements that were enhanced using JS would usually also work with JavaScript disabled in the browser. In some cases, we were able to delete certain legacy behaviors altogether instead of having to rewrite them in vanilla JS.</p>
</li>
</ul>
<p>With these and similar efforts combined over the years, we were able gradually reduce our dependence on jQuery until there was not a single line of code referencing it anymore.</p>
<h2>Custom Elements</h2>
<p>One technology that has been making waves in the recent years is <a href="https://developers.google.com/web/fundamentals/web-components/customelements">Custom Elements</a>: a component library native to the browser, which means that there are no additional bytes of a framework for the user to download, parse and compile.</p>
<p>We had created a few Custom Elements based on the v0 specification since 2014. However, as standards were still in flux back then, we did not invest as much. It was not until 2017 when the Web Components v1 spec was released and implemented in both Chrome and Safari that <a href="https://github.com/search?q=topic%3Aweb-components+org%3Agithub">we began to adopt Custom Elements</a> on a wider scale.</p>
<p>During the jQuery migration, we looked for patterns that would be suitable for extraction as custom elements. For example, we converted our facebox usage for displaying modal dialogs to the <a href="https://github.com/github/details-dialog-element#readme"><code>&lt;details-dialog&gt;</code> element</a>.</p>
<p>Our general philosophy of striving for progressive enhancement extends to custom elements as well. This means that we keep as much of the content in markup as possible and only add behaviors on top of that. For example, <code>&lt;local-time&gt;</code> shows the raw timestamp by default and gets upgraded to translate the time to the local timezone, while <code>&lt;details-dialog&gt;</code>, when nested in the <code>&lt;details&gt;</code> element, is interactive even without JavaScript, but gets upgraded with accessibility enhancements.</p>
<p>Here is an example of how a <code>&lt;local-time&gt;</code> custom element could be implemented:</p>
<div>
<div>
<pre>
<code><span>// The local-time element displays time in the user's current timezone</span>
<span>// and locale.</span>
<span>//</span>
<span>// Example:</span>
<span>// &lt;local-time datetime="2018-09-06T08:22:49Z"&gt;Sep 6, 2018&lt;/local-time&gt;</span>
<span>//</span>
<span>class</span> <span>LocalTimeElement</span> <span>extends</span> <span>HTMLElement</span> <span>{</span>
  <span>static</span> <span>get</span> <span>observedAttributes</span><span>()</span> <span>{</span>
    <span>return</span> <span>[</span><span>'datetime'</span><span>]</span>
  <span>}</span>

  <span>attributeChangedCallback</span><span>(</span><span>attrName</span><span>,</span> <span>oldValue</span><span>,</span> <span>newValue</span><span>)</span> <span>{</span>
    <span>if</span> <span>(</span><span>attrName</span> <span>===</span> <span>'datetime'</span><span>)</span> <span>{</span>
      <span>const</span> <span>date</span> <span>=</span> <span>new</span> <span>Date</span><span>(</span><span>newValue</span><span>)</span>
      <span>this</span><span>.</span><span>textContent</span> <span>=</span> <span>date</span><span>.</span><span>toLocaleString</span><span>()</span>
    <span>}</span>
  <span>}</span>
<span>}</span>

<span>if</span> <span>(</span><span>!</span><span>window</span><span>.</span><span>customElements</span><span>.</span><span>get</span><span>(</span><span>'local-time'</span><span>))</span> <span>{</span>
  <span>window</span><span>.</span><span>LocalTimeElement</span> <span>=</span> <span>LocalTimeElement</span>
  <span>window</span><span>.</span><span>customElements</span><span>.</span><span>define</span><span>(</span><span>'local-time'</span><span>,</span> <span>LocalTimeElement</span><span>)</span>
<span>}</span>
</code>
</pre>
</div>
</div>
<p>One aspect of Web Components that we&rsquo;re looking forward to adopting is <a href="https://developers.google.com/web/fundamentals/web-components/shadowdom">Shadow DOM</a>. The powerful nature of Shadow DOM has the potential to unlock a lot of possibilities for the web, but that also makes it harder to polyfill. Because polyfilling it today incurs a performance penalty even for code that manipulates parts of the DOM <em>unrelated</em> to web components, it is unfeasible for us to start using it in production.</p>
<h2>Polyfills</h2>
<p>These are the polyfills that helped us transition to using standard browser features. We try to serve most of these polyfills only when absolutely necessary, i.e. to outdated browsers as part of a separate &ldquo;compatibility&rdquo; JavaScript bundle.</p>
<ul>
<li><a href="https://github.com/github/eventlistener-polyfill#readme">github/eventlistener-polyfill</a></li>
<li><a href="https://github.com/github/fetch#readme">github/fetch</a></li>
<li><a href="https://github.com/github/form-data-entries#readme">github/form-data-entries</a></li>
<li><a href="https://github.com/iamdustan/smoothscroll#readme">iamdustan/smoothscroll</a></li>
<li><a href="https://github.com/javan/details-element-polyfill#readme">javan/details-element-polyfill</a></li>
<li><a href="https://github.com/jonathantneal/closest#readme">jonathantneal/closest</a></li>
<li><a href="https://github.com/kumarharsh/custom-event-polyfill#readme">kumarharsh/custom-event-polyfill</a></li>
<li><a href="https://github.com/marvinhagemeister/request-idle-polyfill#readme">marvinhagemeister/request-idle-polyfill</a></li>
<li><a href="https://github.com/mathiasbynens/Array.from#readme">mathiasbynens/Array.from</a></li>
<li><a href="https://github.com/mathiasbynens/String.prototype.codePointAt#readme">mathiasbynens/String.prototype.codePointAt</a></li>
<li><a href="https://github.com/mathiasbynens/String.prototype.endsWith#readme">mathiasbynens/String.prototype.endsWith</a></li>
<li><a href="https://github.com/mathiasbynens/String.prototype.startsWith#readme">mathiasbynens/String.prototype.startsWith</a></li>
<li><a href="https://github.com/medikoo/es6-symbol#readme">medikoo/es6-symbol</a></li>
<li><a href="https://github.com/nicjansma/usertiming.js#readme">nicjansma/usertiming.js</a></li>
<li><a href="https://github.com/rubennorte/es6-object-assign#readme">rubennorte/es6-object-assign</a></li>
<li><a href="https://github.com/stefanpenner/es6-promise#readme">stefanpenner/es6-promise</a></li>
<li><a href="https://github.com/webcomponents/template#template">webcomponents/template</a></li>
<li><a href="https://github.com/webcomponents/URL#readme">webcomponents/URL</a></li>
<li><a href="https://github.com/webcomponents/webcomponentsjs#readme">webcomponents/webcomponentsjs</a></li>
<li><a href="https://github.com/WebReflection/url-search-params#readme">WebReflection/url-search-params</a></li>
<li><a href="https://github.com/yola/classlist-polyfill#readme">yola/classlist-polyfill</a></li>
</ul>
<br>
<br>
<br>
<br>
&#25216;&#26415;<br>
<br>
via GitHub Engineering https://ift.tt/2iXnnee<br>
<br>
September 7, 2018 at 03:37AM
